mlir-sycl/doc/design/SYCLKernelSubmissionRepresentation.md

```diff
@@ -0,0 +1,510 @@
+submission process where it is guaranteed that range and argument information is
+available.
+
+The `set_kernel`'s symbol reference points to a `gpu.func` with the `kernel`
+attribute, e.g.:
+
+```mlir
+sycl.host.handler.set_kernel %handler -> @device_functions::@_ZTS10KernelName : !llvm.ptr
+```
+
+As an intermediate step to raising `set_kernel`, the SYCL dialect also defines
+the [`sycl.host.get_kernel`](#syclhostget_kernel-syclsyclhostgetkernelop)
+operation, which replaces `llvm.mlir.addressof` operations that reference a
+`gpu.func` with the `kernel` attribute.
+
+### Raising `schedule_kernel`
+
+With the information associated with the `set_nd_range`, `set_captured` and
+`set_kernel` operations, we can finally raise the `schedule_kernel` operation.
+
+The actual submission of the kernel to the scheduler is not part of the CGF,
+hence we anchor the `RaiseScheduleKernel` pattern at the previously raised
+`set_kernel` operation instead. We scan the CGF for `set_nd_range` and
+`set_captured` ops. As we do not support `single_task` launches yet, we expect
+to find exactly one `set_nd_range` operation.
+
+To ensure that complete and type-correct set of `set_captured` ops has been
+found, we interpret the `kernel_signatures` data structure which is emitted by
+the device compiler into the integration header. To find corresponding kernel
+parameter descriptors, we annotate additional variables in the handler class'
+`StoreLambda` method: 
+
+```c++
+// Excerpt from `sycl::handler::StoreLambda`
+...
+auto numParams __SYCL_ANNOTATE(kernel_num_params) = KI::getNumParams();
+auto *paramDesc __SYCL_ANNOTATE(kernel_param_desc) = &KI::getParamDesc(0);
+extractArgsAndReqsFromLambda(reinterpret_cast<char *>(KernelPtr),
+                             numParams, paramDesc, KI::isESIMD());
+MKernelName = KI::getName(); // raised to `sycl.host.handler.set_kernel`
+...
+```
+In the IR, we expect to find the resulting annotation in the predecessor of the
+block containing the `set_kernel` op (due to the inbetween `llvm.invoke` of
+`extractArgsAndReqsFromLambda`).
+
+#### What to do when raising fails
+
+- If no `set_kernel` op is present in the IR, check whether not enough or too
+  much inlining has happened. The raising approach is brittle and may be
+  affected even by minor changes to the IR structure.
+- If a `set_kernel` op is present, but has not been raised further to
+  `schedule_kernel`, check that there is a unique `set_nd_range` in the same
+  function. More likely though, the consistency check regarding the
+  `set_captured` ops has failed:
+  - Is there a continious sequence of `set_captured` ops present in the same
+    function?
+  - Are accessor arguments captured with a SYCL-type attribute? If not, raising
+    of the `host.constructor` ops may have failed or happened after application
+    of the `RaiseSetCaptured` pattern, or the constructor op is in a different
+    function.
+
+## Appendix: Excerpt from ODS documentation
```

Comment:
How do we keep this in sync with the actual tablegen files/implementation?

Reply:
I think we should have a link to the docs instead.

Reply:
I think the main benefit of having a copy of the rendered docs in here is that this document is self-contained and can be viewed on GitHub. We otherwise don't have a stable target to link to, except something like `../../../build/docs/Dialects/SYCLOps.md` which would only work locally and after building `mlir-sycl-doc`.

Given that the ops probably don't change that often, and changes might require updating this doc anyways, would it be acceptable to just drop a comment in `SYCLHostOps.td` to please remember to update the appendix here as well?

Reply:
I usually just refer to the source `.td`. Not sure what @sommerlukas prefers, but I'd rather have people read the original doc source rather than potentially having this outdated.

Reply:
Right now, we do not have many external users that would consume online documentation.

For the development team, reading the `.td` directly or creating the markdown locally is probably fine and we avoid this document becoming stale in the mean-time.

Reply:
Ok, I'll drop the appendix.